File: love_ga_wrapper/love_ga_wrapper/lib/genetic_algorithms/population/PopulationFactory.lua
local class = require("love_ga_wrapper.lib.classes.class")
local extend = require("love_ga_wrapper.lib.classes.extend")
local Population = require("love_ga_wrapper.lib.genetic_algorithms.population.Population")

-- Mutation Funs
local RegularMutationFunction =  require("love_ga_wrapper.lib.genetic_algorithms.mutation_functions.RegularMutationFunction")

-- Selection Funs
local TournamentSelectionFunction = require("love_ga_wrapper.lib.genetic_algorithms.selection_functions.TournamentSelectionFunction")

-- Crossover Funs
local SinglePointCrossoverFunction = require("love_ga_wrapper.lib.genetic_algorithms.crossover_functions.SinglePointCrossoverFunction")

--------------------------------------------------------------------------------------------------------

-- class: PopulationFactory
-- Factory class for creating populations
local PopulationFactory = class(function(self)
end)

-- getClassicPopulation: list(Individuals), Tester, float(0,1), float(0,1) -> Population
-- returns a classic population for genetic algorithms
function PopulationFactory.getClassicPopulation(individuals, tester, mutation_prob, elitism_ratio)
    local population = Population.new(individuals, tester, elitism_ratio)

    local mutation_manager = population:getMutationManager()
    mutation_manager:setMutationFun(RegularMutationFunction.new(mutation_prob))

    -- Tournament selection with tournaments of size 4
    local selection_manager = population:getSelectionManager()
    selection_manager:setSelectionFun(TournamentSelectionFunction.new(4))

    local crossover_manager = population:getCrossoverManager()
    crossover_manager:serCrossoverFun(SinglePointCrossoverFunction.new())

    return population
end

return PopulationFactory

File: lib/genetic_algorithm/population/PopulationFactory.py
from lib.genetic_algorithm.crossover_functions.SinglePointCrossoverFunction import SinglePointCrossoverFunction
from lib.genetic_algorithm.mutation_functions.RegularMutationFunction import RegularMutationFunction
from lib.genetic_algorithm.population.Population import Population
from lib.genetic_algorithm.population.managers.ParallelTesterManager import ParallelTesterManager
from lib.genetic_algorithm.population.managers.TesterManager import TesterManager
from lib.genetic_algorithm.selection_functions.TournamentSelectionFunction import TournamentSelectionFunction


class PopulationFactory:

	@staticmethod
	def get_classic_population(individuals, tester, mutation_prob, elitism_ratio):
		population = Population(individuals, TesterManager(tester), elitism_ratio)

		# Regular Mutation Function
		mutation_manager = population.get_mutation_manager()
		mutation_manager.set_mutation_fun(RegularMutationFunction(mutation_prob))

		# Tournament Selection with tournaments of size 4
		selection_manager = population.get_selection_manager()
		selection_manager.set_selection_fun(TournamentSelectionFunction(4))

		# Single Point Crossover Function
		crossover_manager = population.get_crossover_manager()
		crossover_manager.set_crossover_fun(SinglePointCrossoverFunction())

		return population

	@staticmethod
	def get_classic_parallel_population(individuals, testers, mutation_prob, elitism_ratio):
		parallel_tester_manager = ParallelTesterManager(testers)
		population = Population(individuals, parallel_tester_manager, elitism_ratio)

		# Regular Mutation Function
		mutation_manager = population.get_mutation_manager()
		mutation_manager.set_mutation_fun(RegularMutationFunction(mutation_prob))

		# Tournament Selection with tournaments of size 4
		selection_manager = population.get_selection_manager()
		selection_manager.set_selection_fun(TournamentSelectionFunction(4))

		# Single Point Crossover Function
		crossover_manager = population.get_crossover_manager()
		crossover_manager.set_crossover_fun(SinglePointCrossoverFunction())

		return population